File: library/src/main/java/work/wanghao/library/RxBus.java
package work.wanghao.library;

import rx.Observable;
import rx.subjects.PublishSubject;
import rx.subjects.SerializedSubject;
import rx.subjects.Subject;

/**
 * Create on: 2016-06-03
 * Author: wangh
 * Summary: TODO
 */
public class RxBus {

  private static volatile RxBus mDefaultInstance;

  private final Subject mBus;

  public RxBus() {
    mBus = new SerializedSubject<>(PublishSubject.create());
  }

  public static RxBus getDefault() {
    RxBus rxBus = mDefaultInstance;
    if (mDefaultInstance == null) {
      synchronized (RxBus.class) {
        rxBus = mDefaultInstance;
        if (mDefaultInstance == null) {
          rxBus = new RxBus();
          mDefaultInstance = rxBus;
        }
      }
    }
    return rxBus;
  }

  public void post(Object event) {
    mBus.onNext(event);
  }

  public <T> Observable<T> toObserverable(Class<T> eventType) {
    return mBus.ofType(eventType);
  }
}

File: library/src/main/java/work/wanghao/rxbus2/RxBus.kt
package work.wanghao.rxbus2

import com.jakewharton.rxrelay2.PublishRelay
import com.jakewharton.rxrelay2.Relay
import io.reactivex.BackpressureStrategy
import io.reactivex.Flowable
import io.reactivex.disposables.CompositeDisposable

/**
 * @author doublemine
 *         Created  on 2017/02/07 15:17.
 *         Summary:
 */

class RxBus private constructor() {

  companion object {
    fun get(): RxBus {
      return Inner.single
    }
  }

  private object Inner {
    val single = RxBus()
  }

  private val mBus: Relay<Any> = PublishRelay.create<Any>().toSerialized()

  fun post(event: Any) {
    mBus.accept(event)
  }

  fun asFlowable(): Flowable<Any> {
    return mBus.toFlowable(BackpressureStrategy.LATEST)
  }

  fun hasObservers(): Boolean {
    return mBus.hasObservers()
  }

  fun asFlowableFilterType(obj: Class<*>): Flowable<*> {
    return mBus.toFlowable(BackpressureStrategy.LATEST).ofType(obj)
  }

  fun register(any: Any) {
    val compositeDisposable = CompositeDisposable()

  }

  fun unRegister(any: Any) {

  }

}